pkg/providers/pricing/pricing.go

```diff
@@ -320,74 +325,66 @@ func (p *DefaultProvider) onDemandPage(ctx context.Context, prices map[string]fl
+		currency = "CNY"
+	}
+	for _, outer := range output.PriceList {
+		pItem := &priceItem{}
+		if err := json.Unmarshal([]byte(outer), pItem); err != nil {
+			log.FromContext(ctx).Error(err, "failed unmarshaling pricing data")
 		}
-		for _, outer := range output.PriceList {
-			var buf bytes.Buffer
-			enc := json.NewEncoder(&buf)
-			if err := enc.Encode(outer); err != nil {
-				log.FromContext(ctx).Error(err, "failed encoding pricing data")
-			}
-			dec := json.NewDecoder(&buf)
-			var pItem priceItem
-			if err := dec.Decode(&pItem); err != nil {
-				log.FromContext(ctx).Error(err, "failed decoding pricing data")
-			}
-			if pItem.Product.Attributes.InstanceType == "" {
-				continue
-			}
-			for _, term := range pItem.Terms.OnDemand {
-				for _, v := range term.PriceDimensions {
-					price, err := strconv.ParseFloat(v.PricePerUnit[currency], 64)
-					if err != nil || price == 0 {
-						continue
-					}
-					prices[pItem.Product.Attributes.InstanceType] = price
+		if pItem.Product.Attributes.InstanceType == "" {
+			continue
+		}
+		for _, term := range pItem.Terms.OnDemand {
+			for _, v := range term.PriceDimensions {
+				price, err := strconv.ParseFloat(v.PricePerUnit[currency], 64)
+				if err != nil || price == 0 {
+					continue
 				}
+				result[ec2types.InstanceType(pItem.Product.Attributes.InstanceType)] = price
 			}
 		}
-		return true
 	}
+
+	return result
 }
 
 // nolint: gocyclo
 func (p *DefaultProvider) UpdateSpotPricing(ctx context.Context) error {
-	prices := map[string]map[string]float64{}
+	prices := map[ec2types.InstanceType]zonal{}
 
 	p.muSpot.Lock()
 	defer p.muSpot.Unlock()
-	err := p.ec2.DescribeSpotPriceHistoryPagesWithContext(
-		ctx,
-		&ec2.DescribeSpotPriceHistoryInput{
-			ProductDescriptions: []*string{
-				aws.String("Linux/UNIX"),
-				aws.String("Linux/UNIX (Amazon VPC)"),
-			},
-			// get the latest spot price for each instance type
-			StartTime: aws.Time(time.Now()),
+
+	input := &ec2.DescribeSpotPriceHistoryInput{
+		ProductDescriptions: []string{
+			"Linux/UNIX",
+			"Linux/UNIX (Amazon VPC)",
 		},
-		p.spotPage(ctx, prices),
-	)
+		// get the latest spot price for each instance type
+		StartTime: aws.Time(time.Now()),
+	}
 
-	if err != nil {
-		return fmt.Errorf("retrieving spot pricing data, %w", err)
+	paginator := ec2.NewDescribeSpotPriceHistoryPaginator(p.ec2, input)
+	for paginator.HasMorePages() {
+		output, err := paginator.NextPage(ctx)
+		if err != nil {
+			return fmt.Errorf("retrieving spot pricing data, %w", err)
+		}
+		prices = lo.Assign(prices, p.spotPage(ctx, output))
 	}
 	if len(prices) == 0 {
 		return fmt.Errorf("no spot pricing found")
 	}
 
 	totalOfferings := 0
 	for it, zoneData := range prices {
-		if _, ok := p.spotPrices[it]; !ok {
-			p.spotPrices[it] = newZonalPricing(0)
-		}
-		for zone, price := range zoneData {
-			p.spotPrices[it].prices[zone] = price
-		}
-		totalOfferings += len(zoneData)
+		p.spotPrices[it] = newZonalPricing(0)
```

Comment:
This is wrong, right? It overrides the previous pricing. Before we only overrode the pricing data when we got new price information.